Use the whole TTableDescription to describe an indexImplTable in TIndexDescription

jepett0 opened the pull request:
It would be better (more future proof) to keep all the necessary settings of the indexImplTable in one message. I would not like to reinvent the bicycle, so the best option I see here is to reuse TTableDescription.

Additionally, I have added options `fillConfig` and `fillBoundaries` to control what parts of the whole TTableDescription would be return on DescribeTableIndex request. It should prevent the unnecessary use of CPU for parsing big protobufs (TableBoundaries can be big). We would need to enable `fillBoundaries` explicitly for backups, but this would be done in scope of https://github.com/ydb-platform/ydb/issues/4955.

And the last change is to make IndexImplTableDescription field a repeated field. The motivation is to prepare `flat_scheme_op.proto` for multiple indexImplTables which are said to be coming soon to implement vector indices. It is a breaking change, but there are no instances of YDB that have been using the new TIndexDescriptions with PartitionConfig and PartitionBoundaries, so it should be ok.

jepett0 committed:
Prepare for multiple indexImplTables in flat_scheme_op.proto

jepett0 commented on ydb/core/grpc_services/rpc_describe_table.cpp:
```diff
@@ -92,7 +92,14 @@ class TDescribeTableRPC : public TRpcSchemeRequestActor<TDescribeTableRPC, TEvDe
 
-                FillIndexDescription(describeTableResult, tableDescription, splitKeyType);
+                try {
+                    FillIndexDescription(describeTableResult, tableDescription);
```
FillIndexDescription now calls FillColumnDescription which might throw

jepett0 commented on ydb/core/tx/schemeshard/schemeshard_path_describer.cpp:
```diff
@@ -1238,12 +1238,7 @@ void TSchemeShard::DescribeTableIndex(const TPathId& pathId, const TString& name
-    ) {
-        *entry.MutableExplicitPartitions()->MutableSplitBoundary() = explicitPartitions;
-    }
+    *entry.AddIndexImplTableDescription() = indexImplTable->TableDescription;
```
The entire TableDescription is copied here in the TIndexDescription. However, it does not contain a lot of info. Mainly, it does not contain TTablePartition which is heavy for multishard tables.

<img width="1728" alt="Screenshot 2024-07-04 at 12 12 30" src="https://github.com/ydb-platform/ydb/assets/111313089/99e21911-fa28-4cd5-986a-6117bdce7d4d">

jepett0 replied:
I wonder why there is no column info in the IndexImplTableDescription... I will solve this problem in scope of https://github.com/ydb-platform/ydb/issues/4955

jepett0 commented on ydb/core/ydb_convert/table_description.cpp:
```diff
@@ -719,10 +719,9 @@ bool FillColumnDescription(NKikimrSchemeOp::TColumnTableDescription& out,
-    const NKikimrMiniKQL::TType& splitKeyType
-) {
-    for (const auto& boundary : boundaries) {
+        const NKikimrSchemeOp::TTableDescription& in, const NKikimrMiniKQL::TType& splitKeyType) {
```
These changes restore these lines to the original state in which they were before the https://github.com/ydb-platform/ydb/pull/5119

jepett0 commented on ydb/core/ydb_convert/table_description.cpp:
```diff
@@ -802,38 +801,53 @@ void FillPartitioningSettings(TYdbProto& out, const NKikimrSchemeOp::TPartitioni
+    if (indexImplTableDescription.HasUniformPartitionsCount()) {
+        settings.set_uniform_partitions(indexImplTableDescription.GetUniformPartitionsCount());
+    }
+    if (indexImplTableDescription.SplitBoundarySize()) {
```
You could ask yourself: "What if __BOTH__ UniformPartitionsCount and SplitBoundary are specified?" My answer is: SplitBoundary has higher priority. The other option I see here is to throw an error. (FillIndexDescription already throws if FillColumnDescription cannot decide how to fill the columns.)

MBkkt replied:
> The other option I see here is to throw an error. (FillIndexDescription already throws if FillColumnDescription cannot decide how to fill the columns.)

Exception sounds better for me, because 
1. looks like error case
2. can be changed to something else in future

jepett0 replied:
I checked how DescribeTable deals with this problem and found out that it just ignores UniformPartitionsCount, i.e. it does not even check if it exists in the TTableDescription. See the code around [here](https://github.com/ydb-platform/ydb/blob/29e1e4546446acf94c2a350d97b3dbfdf68f766f/ydb/core/tx/schemeshard/schemeshard_path_describer.cpp#L1183). The motivation is: there is no such thing as UniformPartitionsCount for an __existing__ table. If a table already exists, then it has __explicit__ partition boundaries which are represented by SplitBoundary field of the TTableDescription protobuf.

azevaykin commented on ydb/core/protos/flat_scheme_op.proto:
```diff
@@ -1000,15 +996,10 @@ message TIndexDescription {
-        TExplicitPartitions ExplicitPartitions = 11;
-    }
-    optional TPartitioningPolicy PartitioningPolicy = 12;
+    repeated TTableDescription IndexImplTableDescription = 10;
```
According to the [Language Guide (proto 3)](https://protobuf.dev/programming-guides/proto3/) , repeated fields are used in the plural.
```suggestion
    repeated TTableDescription IndexImplTableDescriptions = 10;
```

jepett0 replied:
Ok. I agree. More on the issue:
- https://protobuf.dev/programming-guides/style/#repeated-fields
- https://google.aip.dev/144#:~:text=Repeated%20fields%20must%20use%20a,coin%20a%20new%20plural%20form.
- statistically speaking, only 12 out of 94 repeated protobuf field names are singular in flat_scheme_op.proto

jepett0 committed:
review fix

- use plural form for a repeated field name

MBkkt commented on ydb/core/tx/datashard/export_common.cpp:
```diff
@@ -59,7 +59,12 @@ TMaybe<Ydb::Table::CreateTableRequest> GenYdbScheme(
     }
 
-    FillIndexDescription(scheme, tableDesc, mkqlKeyType);
+    try {
+        FillIndexDescription(scheme, tableDesc);
+    } catch (const yexception&) {
+        return Nothing();
+    }
```
NIT: There is no need to separate try catch for FillTableBoundary and FillIndexDescription

jepett0 commented on ydb/core/protos/flat_scheme_op.proto:
```diff
@@ -1000,15 +996,10 @@ message TIndexDescription {
 
     repeated string DataColumnNames = 8;
+
     // DataSize + IndexSize of indexImplTable
```
IndexSize is the size of the index of the indexImplTable. It follows the general DataSize + IndexSize = TotalSize [convention](https://github.com/ydb-platform/ydb/blob/6f872bdf6fdd00795d70a46fb00121bb84d808e4/ydb/core/protos/table_stats.proto#L32) for tables.

jepett0 replied:
I would much rather leave this comment for you to deal with 😜

jepett0 committed:
Refactoring

- replace contains + at with a single FindPtr
- rename variables to be shorter

jepett0 committed:
Add fillConfig and fillBoundaries options for DescribeTableIndex

Just like in DescribeTable

jepett0 committed:
Review fix

- UniformPartitionsCount makes sense only for CreateTable operation, not DescribeTable. If you look into DescribeTable, it does not bother checking uniform partitions count at all

jepett0 commented on ydb/core/tx/schemeshard/schemeshard_impl.h:
```diff
@@ -1016,8 +1016,12 @@ class TSchemeShard
-    void DescribeTableIndex(const TPathId& pathId, const TString& name, NKikimrSchemeOp::TIndexDescription& entry);
-    void DescribeTableIndex(const TPathId& pathId, const TString& name, TTableIndexInfo::TPtr indexInfo, NKikimrSchemeOp::TIndexDescription& entry);
+    void DescribeTableIndex(const TPathId& pathId, const TString& name,
+        bool fillConfig, bool fillBoundaries, NKikimrSchemeOp::TIndexDescription& entry
```
I have added explicit `fillConfig` and `fillBoundaries` options to be able to control the behavior of `DescribeTableIndex`. Moreover, these options are supported by `DescribeTable` and it would be inconsistent to ignore them in `DescribeTableIndex`

jepett0 commented on ydb/core/tx/schemeshard/schemeshard_path_describer.cpp:
```diff
@@ -8,6 +8,16 @@
+void FillPartitionConfig(const NKikimrSchemeOp::TPartitionConfig& in, NKikimrSchemeOp::TPartitionConfig& out) {
+    out.CopyFrom(in);
+    NKikimr::NSchemeShard::TPartitionConfigMerger::DeduplicateColumnFamiliesById(out);
+    out.MutableStorageRooms()->Clear();
```
It turns out that PartitionConfig is not just copied, but some unnecessary info is erased before giving it out to users

jepett0 committed:
Review fix

- do not blatantly use returnBoundaries option from the parent table in DescribeTableIndex
